tc/autotuner/autotuner.h

```diff
@@ -0,0 +1,214 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <atomic>
+#include <csignal>
+#include <deque>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "tc/aten/aten.h"
+#include "tc/aten/aten_compiler.h"
+
+#include "tc/autotuner/genetic_search.h"
+#include "tc/autotuner/options_cache.h"
+#include "tc/autotuner/parameters.h"
+#include "tc/autotuner/utils.h"
+#include "tc/core/tensor.h"
+#include "tc/core/utils/time.h"
+#include "tc/lang/parser.h"
+
+namespace tc {
+namespace autotune {
+
+namespace detail {
+/**
+ * Internal harness to support multithreaded compilation and evaluation over
+ * various backends for a particular SearchStrategy.
+ * The SearchStrategy is a template parameter passed to the run method.
+ */
+template <typename Backend>
+class TuningHarness {
+ public:
+  using BackendType = Backend;
+  using ExecutorType = typename BackendType::ExecutorType;
+  using MappingOptionsType = typename BackendType::MappingOptionsType;
+  using OptionsCacheType = OptionsCache<Backend>;
+
+  TuningHarness(
+      size_t maxPopulationSize,
+      lang::TreeRef kTcTree,
+      const std::unordered_map<size_t, std::vector<const DLConstTensor*>>&
+          inputs,
+      std::unordered_map<size_t, std::vector<const DLTensor*>>& outputs,
+      const typename Backend::MappingOptionsType& baseMapping,
+      const TuningParameterFixer& fixedParams,
+      std::shared_ptr<OptionsCache<Backend>>& optionsCache);
+
+  /// Runs a SearchStrategy over numIterations
+  template <typename SearchStrategy>
+  void run(SearchStrategy* searchStrategy);
+
+  /// Once a signal has been caught, a flag is set and we terminate after the
+  /// current iteration is finished. This is apparently necessary for proper
+  /// python termination.
+  /// TODO: we should detect when we come from python and exit properly in C++.
+  void stopAfterCurrentIteration();
+
+ private:
+  /// Traverse one iteration of candidates in parallel and evaluate their
+  /// runtimes
+  template <typename SearchStrategy>
+  void runOneIteration(SearchStrategy* searchStrategy, size_t iteration);
+
+  /// Helper function to delegate compiling on the cpu to different threads
+  template <typename SearchStrategy>
+  void doCompile(SearchStrategy* searchStrategy);
+
+  /// Helper function to get a kernel into benchmark-able state
+  /// This function must be specialized per Backend
+  /// TODO: virtual inheritance?
```

Comment:
What's the alternative?

Reply:
Renamed to ```skipExecution``` and documenting it.
In the current impl, the function is not defined in the template case.
It is specialized per backend, like:
```
using CudaTuningHarness = class TuningHarness<tc::CudaBackend>;

template <>
bool CudaTuningHarness::skipExecution(...) {}
```

This is not good because it results in link-time errors when a backend autotuner does not implement it.

This was mostly to avoid .h proliferation but it's too bad. I'll make proper derived CudaTuningHarness and use a pure virtual function in the base instead.

Reply:
An alternative would be to stick that function as a static member class in the ```Backend``` type which is maybe better. Actually this function is of the same category as the ones at the bottom of the file.

```makeOptions```, ```makeTuningConfiguration```, ```handleDeviceRuntimeError``` and ```skipExecutionOrWarmup``` are all Backend-specific behaviors I haven't really found a good place for.

Note these are Backend-specific but only used in tuning. Probably ```makeOptions``` and ```makeTuningConfiguration``` can go away if the ```TuningConfiguration``` and ```parameters``` are templated properly and made backend-specific too.

In the meantime I'll actually extract ```skipExecutionOrWarmup``` and put it close to the other functions where it belongs.